core/server/common/src/main/java/alluxio/master/PrimarySelectorClient.java

```diff
@@ -132,7 +139,54 @@ public void start(InetSocketAddress address) throws IOException {
 
   @Override
   public void stateChanged(CuratorFramework client, ConnectionState newState) {
-    setState(State.SECONDARY);
+    // Handle state change.
+    switch (newState) {
```

Comment:
This seems difficult to follow and maintain. I don't know if it makes sense to make subclasses, but maybe we can create 2 methods to handle the state change, one for each policy, and just do a single switch here on the policy?

Reply:
Agreed.